File: android/src/main/java/com/aiweiju/azure_sdk_voice/AzureSdkVoicePlugin.java
package com.aiweiju.azure_sdk_voice;

import androidx.annotation.NonNull;

import android.content.Context;
import android.util.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** AzureSdkVoicePlugin */
public class AzureSdkVoicePlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private Context context;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "azure_sdk_voice");
    channel.setMethodCallHandler(this);

    context = flutterPluginBinding.getApplicationContext();
    recorderPlayer = new AudioRecorderPlayer(context);
  }

  private String key;
  private String region;

  private AudioRecorderPlayer recorderPlayer;

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("init")) {
      Log.d("flutter_spf", "init");
      key = call.argument("key");
      region = call.argument("region");
      result.success("OK");
    } else if (call.method.equals("startRecording")) {
      Log.d("flutter_spf", "startRecording");
      String fileName = call.argument("name");
      recorderPlayer.startRecording(fileName);
      result.success("OK");
    } else if (call.method.equals("stopRecording")) {
      Log.d("flutter_spf", "stopRecording");
      recorderPlayer.stopRecording();
      result.success("OK");
    } else if (call.method.equals("playRecordedAudio")) {
      Log.d("flutter_spf", "playRecordedAudio");
      String fileName = call.argument("name");
      recorderPlayer.startPlaying(fileName);
      result.success("OK");
    } else if (call.method.equals("pronunciationScore")) {
      Log.d("flutter_spf", "pronunciationScore");
      String fileName = call.argument("name");
      String language = call.argument("language");
      String topic = call.argument("topic");

      String text = AzureSdk.instance.pronunciationAssessmentWithContentAssessment(key,region,fileName,language,topic);
      result.success(text);
    } else if (call.method.equals("translate")) {
      Log.d("flutter_spf", "translate");
      String fileName = call.argument("name");
      String recLanguage = call.argument("recLanguage");
      String toLanguage = call.argument("toLanguage");
      try {
        String text = AzureSdk.instance.translateWav(key,region,fileName,recLanguage,toLanguage);
        result.success(text);
        return;
      } catch (InterruptedException e) {
        e.printStackTrace();
      } catch (ExecutionException e) {
        e.printStackTrace();
      }
      result.success("error");
    } else if (call.method.equals("speak")) {
      Log.d("flutter_spf", "speak");
      String content = call.argument("content");
      int code = call.argument("code");
      try {
        AzureSdk.instance.synthesisToSpeaker(key,region,content,new Callback() {
          @Override
          public void onCallback(String result,boolean is_last) {
            Map<String, Object> arguments = new HashMap<>();
            arguments.put("code", code);
            arguments.put("res", result);
            arguments.put("is_last", is_last);
            channel.invokeMethod("callback", arguments);
          }
        });
      } catch (InterruptedException e) {
        e.printStackTrace();
      } catch (ExecutionException e) {
        e.printStackTrace();
      }
      result.success("OK");
    } else if (call.method.equals("speak_stop")) {
      Log.d("flutter_spf", "speak_stop");
      AzureSdk.instance.stopSynthesisToSpeaker();
      result.success("OK");
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}

File: ios/Classes/AzureSdkVoicePlugin.swift
import Flutter
import UIKit

public class AzureSdkVoicePlugin: NSObject, FlutterPlugin {
    static var gChannel: FlutterMethodChannel?;
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "azure_sdk_voice", binaryMessenger: registrar.messenger())
      gChannel = channel
    let instance = AzureSdkVoicePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  var _audioRecorder: AudioRecorder = AudioRecorder();
  var _azureSdk: AzureSdk = AzureSdk();

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
      guard let arguments = call.arguments as? [String: Any],
            let keyTmp = arguments["key"] as? String,
            let regionTmp = arguments["region"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
        _azureSdk.initKeyRegion(key: keyTmp, region: regionTmp)
      result("OK")
    case "startRecording":
      guard let arguments = call.arguments as? [String: Any],
            let nameTmp = arguments["name"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
      _audioRecorder.startRecording(fileName: nameTmp);
      result("OK")
    case "stopRecording":
      _audioRecorder.stopRecording();
      result("OK")
    case "playRecordedAudio":
      guard let arguments = call.arguments as? [String: Any],
            let nameTmp = arguments["name"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
      _audioRecorder.playRecordedAudio(recordedFileName: nameTmp)
      result("OK")
    case "pronunciationScore":
      guard let arguments = call.arguments as? [String: Any],
            let languageTmp = arguments["language"] as? String,
            let topicTmp = arguments["topic"] as? String,
            let nameTmp = arguments["name"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
        _azureSdk.pronunciationAssessmentWithContentAssessment(fileName: nameTmp,language: languageTmp,topic: topicTmp) {res in
        result(res)
      }
    case "translate":
      guard let arguments = call.arguments as? [String: Any],
            let recLanguageTmp = arguments["recLanguage"] as? String,
            let toLanguageTmp = arguments["toLanguage"] as? String,
            let nameTmp = arguments["name"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
        _azureSdk.translateWav(fileName: nameTmp,recLanguage: recLanguageTmp,toLanguage: toLanguageTmp) {res in
        result(res)
      }
    case "startTranslateContinuous":
        guard let arguments = call.arguments as? [String: Any],
              let recLanguageTmp = arguments["recLanguage"] as? String,
              let toLanguageTmp = arguments["toLanguage"] as? String,
              let code = arguments["code"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        _azureSdk.startTranslateContinuous(recLanguage: recLanguageTmp,toLanguage: toLanguageTmp) {res,is_last in
            DispatchQueue.global(qos: .userInitiated).async {
                let arg2:[String:Any] = ["code":code,"res": res!,"is_last":is_last!]
                AzureSdkVoicePlugin.gChannel?.invokeMethod("callback", arguments: arg2)
            }
        }
        result("OK")
    case "stopTranslateContinuous":
        _azureSdk.stopTranslateContinuous()
        result("OK")
    case "speak":
      guard let arguments = call.arguments as? [String: Any],
            let code = arguments["code"] as? Int,
            let contentTmp = arguments["content"] as? String else {
          result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
          return
      }
        _azureSdk.stopSynthesisToSpeaker()
        DispatchQueue.global(qos: .userInitiated).async {
            let arg2:[String:Any] = ["code":code,"res": "start","is_last":false]
            AzureSdkVoicePlugin.gChannel?.invokeMethod("callback", arguments: arg2)
        
        
            self._azureSdk.synthesisToSpeaker(inputText: contentTmp)
        

            let arg3:[String:Any] = ["code":code,"res": "stop","is_last":true]
            AzureSdkVoicePlugin.gChannel?.invokeMethod("callback", arguments: arg3)
        }
        
        
        result("OK")
    case "speak_stop":
        _azureSdk.stopSynthesisToSpeaker()
        result("OK")
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
